### PR TITLE
capsule for sandbox docker container and vagrant vm 

### PR DESCRIPTION
**Pull request by zeeshanlakhani**
- [x] submod change
- [x] uncomment d.pull for images once we can build them
- [x] name clarity throughout

### CODE REVIEW

File: docker.mk [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        --build-arg RUST_BASE_IMG=$(RUST_BASE_IMG) \
+        --build-arg RR_VERSION=$(RR_VERSION) \
+		-t $(SANDBOX) $(BASE_DIR)
+
+build-all: build-dpdk build-devbind build-mod build-sandbox
+
+connect:
+	@docker exec -it $(SANDBOX_IMG) /bin/bash
+
+pull-all: pull pull-devbind pull-mod pull-sandbox
```
Comment (drunkirishcoder) [inline docker.mk]: pull-dpdk

File: docker.mk [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+build-mod:
+	@docker build --target $(DPDK_MOD_IMG) \
+		--build-arg DPDK_VERSION=$(DPDK_VERSION) \
+		-t $(NAMESPACE)/$(DPDK_MOD_IMG):$(DPDK_VERSION)-$(DPDK_MOD_KERNEL) $(BASE_DIR)
+
+build-sandbox:
+	@docker build --target $(SANDBOX_IMG) \
+		--build-arg DPDK_VERSION=$(DPDK_VERSION) \
+        --build-arg RUST_BASE_IMG=$(RUST_BASE_IMG) \
```
Comment (drunkirishcoder) [inline docker.mk]: not tabs.

File: docker.mk [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+push-sandbox-latest:
+	@docker tag $(SANDBOX) $(SANDBOX_LATEST)
+	@docker push $(SANDBOX_LATEST)
+
+run:
+	@if [ "$$(docker images -q $(SANDBOX))" = "" ]; then \
+		docker pull $(SANDBOX); \
+	fi
+	@docker run -it --rm --privileged --network=host --name $(SANDBOX_IMG) \
+    --cap-add=SYS_PTRACE --security-opt seccomp=unconfined \
```
Comment (drunkirishcoder) [inline docker.mk]: tabs again.

File: docker.mk [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    -w /home/$(NAMESPACE) \
+	-v /lib/modules:/lib/modules \
+	-v /dev/hugepages:/dev/hugepages \
+	-v $(BASE_DIR)/nb2:/home/$(NAMESPACE) \
```
Comment (drunkirishcoder) [inline docker.mk]: change the mount

File: scripts/setup.sh [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    libnuma-dev \
+    libpcap-dev \
+    libssl-dev \
+    linux-headers-$(uname -r) \
+    llvm-dev \
+    pkg-config \
+    python3-setuptools \
+    python3-pip \
+    wget \
+  && sudo pip3 install --system wheel \
```
Comment (drunkirishcoder) [inline scripts/setup.sh]: maybe new line `wheel`.

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+  && apt-get install -y \
+    build-essential \
+    libnuma-dev \
+    libpcap-dev \
+    linux-headers-$(uname -r) \
+    python3-setuptools \
+    python3-pip \
+    wget \
+    ninja-build \
+  && pip3 install wheel \
```
Comment (drunkirishcoder) [inline Dockerfile]: maybe new line `wheel`.

File: scripts/vm.sh [hunk 7]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env bash
+
+DPDK_DRIVER=${1:-uio_pci_generic}
+DPDK_DEVICES=${2:-0000:00:08.0}
+VAGRANT_HOME_DIR=${3:-/home/vagrant}
```
Comment (drunkirishcoder) [inline scripts/vm.sh]: `HOME_DIR`? since i assume this can be used on bare linux install too.

File: scripts/vm.sh [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+#   - cargo-expand (https://crates.io/crates/cargo-expand).
+cargo install cargo-watch && cp $HOME/.cargo/bin/cargo-watch ${VAGRANT_HOME_DIR}/.cargo/bin
+cargo install cargo-expand && cp $HOME/.cargo/bin/cargo-expand ${VAGRANT_HOME_DIR}/.cargo/bin
+
+# devbind to default device
+dpdk-devbind.py --force -b ${DPDK_DRIVER} ${DPDK_DEVICES}
+
+# insmod kni
+if lsmod | grep rte_kni &> /dev/null ; then
+	echo "rte_knii loaded"
```
Comment (drunkirishcoder) [inline scripts/vm.sh]: I think we can just silently ignore rather than echo.

File: Dockerfile [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+LABEL maintainer="williamofockham <occam_engineering@comcast.com>"
+
+COPY --from=builder /usr/local/lib/x86_64-linux-gnu /usr/local/lib/x86_64-linux-gnu
+
+RUN apt-get update \
+  && apt-get upgrade -y \
+  && apt-get install -y \
+    libnuma-dev \
+    libpcap-dev \
+  && rm -rf /var/lib/apt/lists /var/cache/apt/archives
```
[the remark this answers is not in the document]
Comment (zeeshanlakhani) [reply Dockerfile]: done.

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+ARG DPDK_TARGET=/usr/local/src/dpdk-stable-${DPDK_VERSION}
+ARG RR_VERSION
+
+ENV CARGO_INCREMENTAL=0
+ENV RUST_BACKTRACE=1
+
+COPY --from=builder /usr/local/bin /usr/local/bin
+COPY --from=builder /usr/local/lib/x86_64-linux-gnu /usr/local/lib/x86_64-linux-gnu
+COPY --from=builder /usr/local/include /usr/local/include
+COPY --from=builder ${DPDK_TARGET} ${DPDK_TARGET}
```
[the remark this answers is not in the document]
Comment (zeeshanlakhani) [reply Dockerfile]: done did earlier.

File: Dockerfile [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+RUN apt-get update \
+  && apt-get upgrade -y \
+  && apt-get install -y \
+    build-essential \
+    ca-certificates \
+    clang \
+    cpufrequtils \
```
[the remark this answers is not in the document]
Comment (zeeshanlakhani) [reply Dockerfile]: really good for running on vms and making things faster.

File: scripts/vm.sh (hunk 7 above)
Comment (drunkirishcoder) [reply scripts/vm.sh]: no change needed after discussion.

File: Dockerfile (hunk 11 above)
[the remark this answers is not in the document]
Comment (zeeshanlakhani) [reply Dockerfile]: removed though, as per discussion.

File: Vagrantfile [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+  end
+end
+
+# Default Vagrant vars.
+$devbind_img = "capsule/dpdk-devbind:18.11.6"
+$dpdkmod_img = "capsule/dpdk-mod:18.11.6-`uname -r`"
+$sandbox_img = "capsule/sandbox:18.11.6-1.42"
+
+$dpdk_driver = "uio_pci_generic"
+$dpdk_devices = "0000:00:08.0"
```
Comment (drunkirishcoder) [inline Vagrantfile]: we need a second interface. the `nat64` example uses 2.

```
[[ports]]
    name = "eth1"
    device = "0000:00:08.0"
    cores = [1]

[[ports]]
    name = "eth2"
    device = "0000:00:09.0"
    cores = [1]
```

File: scripts/vm-vagrant.sh [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+DPDK_DRIVER=${1:-uio_pci_generic}
+DPDK_DEVICES=${2:-0000:00:08.0}
+VAGRANT_HOME_DIR=${3:-/home/vagrant}
+
+# `modprobe` DPDK driver.
+modprobe $DPDK_DRIVER
+
+# Set vm hugepages for sysctl.
+grep -qxF "vm.nr_hugepages = 2048" /etc/sysctl.conf || \
```
Comment (drunkirishcoder) [inline scripts/vm-vagrant.sh]: this is what I was asking earlier, wondering is it "better" to use `if; then fi` or `||` for these lines.

File: Vagrantfile (hunk 12 above)
Comment (zeeshanlakhani) [reply Vagrantfile]: Done did.

File: scripts/vm-vagrant.sh (hunk 13 above)
Comment (zeeshanlakhani) [reply scripts/vm-vagrant.sh]: @drunkirishcoder ill ask on slack.

File: .github/ISSUE_TEMPLATE/bug-report.md [hunk 14]
```diff
@@ -0,0 +1,22 @@
+---
+name: Bug report
+about: Report a bug in the Capsule sandbox
```
Comment (zeeshanlakhani) [inline .github/ISSUE_TEMPLATE/bug-report.md]: @drunkirishcoder how best to handle sandbox capitalization here?

**Pull request merged**